java/ql/lib/semmle/code/java/frameworks/MyBatis.qll

```diff
@@ -105,94 +105,88 @@ class TypeParam extends Interface {
-      exists(Annotation a, Method m, TypeLiteral type, Class c |
-        a.getType() instanceof MyBatisProvider and
-        type = a.getValue(["type", "value"]) and
-        c.hasMethod(m, type.getTypeName().getType()) and
-        m.hasName(a.getValue("method").(StringLiteral).getValue()) and
-        this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
-      )
-    }
+private class MyBatisAbstractSQLMethodNames extends string {
+  MyBatisAbstractSQLMethodNames() {
+    this in [
+        "SELECT", "OFFSET_ROWS", "FETCH_FIRST_ROWS_ONLY", "OFFSET", "LIMIT", "ORDER_BY", "HAVING",
+        "GROUP_BY", "WHERE", "OUTER_JOIN", "RIGHT_OUTER_JOIN", "LEFT_OUTER_JOIN", "INNER_JOIN",
+        "JOIN", "FROM", "DELETE_FROM", "SELECT_DISTINCT", "SELECT", "INTO_VALUES", "INTO_COLUMNS",
+        "VALUES", "INSERT_INTO", "SET", "UPDATE"
+      ]
   }
+}
 
-  class MyBatisAdditionalTaintStep extends TaintTracking::AdditionalTaintStep {
-    abstract override predicate step(DataFlow::Node node1, DataFlow::Node node2);
+class MyBatisInjectionSink extends DataFlow::Node {
+  MyBatisInjectionSink() {
+    exists(Annotation a, Method m, TypeLiteral type, Class c |
+      a.getType() instanceof MyBatisProvider and
+      type = a.getValue(["type", "value"]) and
+      c.hasMethod(m, type.getTypeName().getType()) and
+      m.hasName(a.getValue("method").(StringLiteral).getValue()) and
+      this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
+    )
   }
+}
 
-  private class MyBatisProviderStep extends MyBatisAdditionalTaintStep {
-    override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
-      exists(
-        MethodAccess ma, Annotation a, Method annotatedMethod, Method providerMethod,
-        TypeLiteral type, Class c
-      |
-        a.getType() instanceof MyBatisProvider and
-        annotatedMethod.getAnAnnotation() = a and
-        ma.getMethod() = annotatedMethod and
-        ma.getAnArgument() = n1.asExpr() and
-        type = a.getValue(["type", "value"]) and
-        providerMethod.hasName(a.getValue("method").(StringLiteral).getValue()) and
-        c.hasMethod(providerMethod, type.getTypeName().getType()) and
-        providerMethod.getAParameter() = n2.asParameter()
-      )
-    }
+private class MyBatisProviderStep extends TaintTracking::AdditionalTaintStep {
+  override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+    exists(
+      MethodAccess ma, Annotation a, Method annotatedMethod, Method providerMethod,
+      TypeLiteral type, Class c
+    |
+      a.getType() instanceof MyBatisProvider and
+      annotatedMethod.getAnAnnotation() = a and
+      ma.getMethod() = annotatedMethod and
+      ma.getAnArgument() = n1.asExpr() and
+      type = a.getValue(["type", "value"]) and
+      providerMethod.hasName(a.getValue("method").(StringLiteral).getValue()) and
+      c.hasMethod(providerMethod, type.getTypeName().getType()) and
+      providerMethod.getAParameter() = n2.asParameter()
+    )
```

Comment:
Similarly. Also, you need to wire arguments with parameters based on the index, otherwise you are creating steps for each argument witch each parameter (arg0 -> param0, arg0 -> param1, arg1 -> param0, arg1 -> param1, and so on):

```suggestion
    exists(MethodAccess ma, Annotation a, Method providerMethod |
      exists(int i |
        ma.getArgument(i) = n1.asExpr() and
        providerMethod.getParameter(i) = n2.asParameter()
      )
    |
      a.getType() instanceof MyBatisProvider and      
      ma.getMethod().getAnAnnotation() = a and
      providerMethod.getDeclaringType() =
        a.getValue(["type", "value"]).(TypeLiteral).getTypeName().getType() and
      providerMethod.hasName(a.getValue("method").(StringLiteral).getValue())
    )
```

Reply:
You are right! Thank you :)